lib/executor/ramping_arrival_rate.go

```diff
@@ -343,13 +356,7 @@ func (varr RampingArrivalRate) Run(parentCtx context.Context, out chan<- stats.S
 	defer close(makeUnplannedVUCh)
 	go func() {
 		defer close(returnedVUs)
-		defer func() {
-			// this is done here as to not have an unplannedVU in the middle of initialization when
-			// starting to return activeVUs
-			for i := uint64(0); i < atomic.LoadUint64(&activeVUsCount); i++ {
-				<-activeVUs
-			}
-		}()
```

Comment:
I understand why this was deleted, but I don't see what replaces it? What is ensuring that all iterations are done before the executor is done and before the VUs are returned to the [common VU buffer pool in the `ExecutionState`](https://github.com/k6io/k6/blob/v0.31.1/lib/execution.go#L177)?

Reply:
Also, this has the effect of ignoring `gracefulStop`, try running the following script with your branch and with the latest k6 `master` (or the official release):
```js
import { sleep } from 'k6';

export let options = {
    scenarios: {
        rarr: {
            executor: 'ramping-arrival-rate',
            startRate: 10, timeUnit: '1s',
            stages: [
                { target: 10, duration: '5s' },
            ],
            preAllocatedVUs: 10,
            maxVUs: 10,
            gracefulStop: '30s',
        },
    }
}

export default function () {
    sleep(10); // longer than the duration, but shorter than duration+gracefulStop
    console.log(`VU ${__VU} finished iteration ${__ITER}`)
}
```

With the k6 `master` code it will wait for the first 10 iterations to finish, because of the `gracefulStop`, and you'd get something like this:
```
  scenarios: (100.00%) 1 scenario, 10 max VUs, 35s max duration (incl. graceful stop):
           * rarr: Up to 10.00 iterations/s for 5s over 1 stages (maxVUs: 10, gracefulStop: 30s)

WARN[0001] Insufficient VUs, reached 10 active VUs and cannot initialize more  executor=ramping-arrival-rate scenario=rarr
INFO[0010] VU 2 finished iteration 0                     source=console
INFO[0010] VU 4 finished iteration 0                     source=console
INFO[0010] VU 3 finished iteration 0                     source=console
INFO[0010] VU 5 finished iteration 0                     source=console
INFO[0010] VU 6 finished iteration 0                     source=console
INFO[0011] VU 7 finished iteration 0                     source=console
INFO[0011] VU 8 finished iteration 0                     source=console
INFO[0011] VU 9 finished iteration 0                     source=console
INFO[0011] VU 10 finished iteration 0                    source=console
INFO[0011] VU 1 finished iteration 0                     source=console

running (11.0s), 00/10 VUs, 10 complete and 0 interrupted iterations
rarr ✓ [======================================] 10/10 VUs  5s  10 iters/s

     data_received........: 0 B 0 B/s
     data_sent............: 0 B 0 B/s
     dropped_iterations...: 39  3.544626/s
     iteration_duration...: avg=10s min=10s med=10s max=10s p(90)=10s p(95)=10s
     iterations...........: 10  0.908878/s
     vus..................: 10  min=10 max=10
     vus_max..............: 10  min=10 max=10
```

with your branch, I get:
```
  scenarios: (100.00%) 1 scenario, 10 max VUs, 35s max duration (incl. graceful stop):
           * rarr: Up to 10.00 iterations/s for 5s over 1 stages (maxVUs: 10, gracefulStop: 30s)

WARN[0001] Insufficient VUs, reached 10 active VUs and cannot initialize more  executor=ramping-arrival-rate scenario=rarr

running (05.0s), 00/10 VUs, 0 complete and 10 interrupted iterations
rarr ✓ [======================================] 08/10 VUs  5s  10 iters/s
WARN[0005] No script iterations finished, consider making the test duration longer 

     data_received........: 0 B 0 B/s
     data_sent............: 0 B 0 B/s
     dropped_iterations...: 39  7.798556/s
     vus..................: 10  min=10 max=10
     vus_max..............: 10  min=10 max=10
```